handout/di/angular2/avoiding_collisions_opaque_token.md

```diff
@@ -0,0 +1,98 @@
+# Avoiding Injection Collisions: OpaqueToken
+
+Since Angular allows the use of tokens as identifiers to its dependency injection system, one of the potential issues is using the same token to represent different entities. If, for example, the string `'token'` is used to inject an entity, it's possible that something totally unrelated also uses `'token'` to inject a different entity. When it comes time for Angular to resolve one of these entities, it might be resolving the wrong one. This behaviour might happen rarely, or are easy to resolve when it happens within a small team, but when it comes to multiple teams working separately on the same codebase or integration of 3rd party modules, the cost of these collisions goes up.
+
+Consider this example where the main app is a consumer of two modules: one that provides an email service and another that provides a logging service.
+
+_app/email/email.service.ts_
+```typescript
+export const apiConfig = 'api-config';
+
+@Injectable()
+export class EmailService {
+  constructor(@Inject(apiConfig) public apiConfig) { }
+}
+```
+
+_app/email/email.module.ts_
+```typescript
+@NgModule({
+  providers: [ EmailService ],
+})
+export class EmailModule { }
+```
+
+The email service has an associated api that is trying to inject configuration settings identified by the string `'api-config'`.
+
+_app/logger/logger.service.ts_
+```typescript
+export const apiConfig = 'api-config';
+
+@Injectable()
+export class LoggerService {
+  constructor(@Inject(apiConfig) public apiConfig) { }
+}
+```
+
+_app/logger/logger.module.ts_
+```typescript
+@NgModule({
+  providers: [ LoggerService ],
+})
+export class LoggerModule { }
+```
+
+`LoggerModule` has an api as well, and it's also using the string `'api-config'` to inject its configuration settings. When it comes time for the main app to specify those settings, Angular overwrites the first `emailApiConfig` value with the `loggerApiConfig` value, since it was provided last.
```

Comment:
My proposal is to move some of your content after showing the `AppModule`. This is my proposed new paragraph:

---

In an effort to avoid a collision between the two tokens with the same name, we could try to rename the imports as shown below.